Refactor types and handling around navigation promise

atscott opened the pull request:
## Commit 1
[refactor(router): fix internal types of resolve and reject promises](https://github.com/angular/angular/commit/deea1b6194ccff6dcbc6bc3993e7e8e13c67f730) 

This commit fixes the internal types for the resolve and reject
functions used for the router navigation Promise. These are currently
typed as `any`.

## commit 2
[refactor(router): Correct resolve value for error handler case](https://github.com/angular/angular/commit/95b616f5ed333289a5bffa810d84978278911db4) 

This commit fixes the types on the error handler resolve and reject
functions and also ensures that the resolve value matches the API type
(`boolean`, not "any random value returned from `errorHandler`"). This
could be considered a breaking change but I would argue instead that
relying on the return value of the error handler is API misuse because
the promise returned by the router navigation _does not_ include `any`
in its type.

## commit 3
[refactor(router): move navigation error handler call to location of error emit](https://github.com/angular/angular/commit/5b1566459df206d21a1d9165d5a88c92eb102796) 

This commit moves the call to the error handler to the location where
the error event emits instead of having it indirectly work through the
events subscription. This change would also allow the handler to return
a value which can be handled by the navigation transition.

## commit 4
[refactor(router): Navigation promise should resolve false instead of `null` on skip](https://github.com/angular/angular/pull/55068/commits/b0a1996739095efad44d987fdb5c5bd54330def5) 

This commit updates the internal resolve value of the navigation promise
to use `false` instead of `null` when a navigation is skipped. The
navigation promise type requires `boolean` so the correct resolution to
match the truthy/falsiness of the currently resolved value, `null`, is
`false` instead.

atscott committed:
refactor(router): fix internal types of resolve and reject promises

This commit fixes the internal types for the resolve and reject
functions used for the router navigation Promise. These are currently
typed as `any`.

atscott committed:
refactor(router): Correct resolve value for error handler case

This commit fixes the types on the error handler resolve and reject
functions and also ensures that the resolve value matches the API type
(`boolean`, not "any random value returned from `errorHandler`"). This
could be considered a breaking change but I would argue instead that
relying on the return value of the error handler is API misuse because
the promise returned by the router navigation _does not_ include `any`
in its type.

atscott committed:
refactor(router): move navigation error handler call to location of error emit

This commit moves the call to the error handler to the location where
the error event emits instead of having it indirectly work through the
events subscription. This change would also allow the handler to return
a value which can be handled by the navigation transition.

AndrewKushnir commented on packages/router/src/navigation_transition.ts:
```diff
@@ -482,7 +494,7 @@ export class NavigationTransitions {
                 ),
               );
-              t.resolve(null);
+              t.resolve(null!);
```
It'd be helpful to have a quick comment in the code to explain why we resolve with `null`, since it doesn't match updated types.

atscott committed:
refactor(router): Navigation promise should resolve `false` instead of `null` on skip

This commit updates the internal resolve value of the navigation promise
to use `false` instead of `null` when a navigation is skipped. The
navigation promise type requires `boolean` so the correct resolution to
match the truthy/falsiness of the currently resolved value, `null`, is
`false` instead.

atscott commented:
This PR was merged into the repository by commit 860de662b6caf840f74dcc985c9bba09eab63e39.
